File: test-dest/22_lambdaLock.swift
public class Lock {
  func lock() {
    print("  locked")
  }
  func unlock() {
    print("  unlocked")
  }
  init() {
  }
}

func lock<T>(lock: Lock, body: () -> T) -> T {
  lock.lock()
  let x = body()
  lock.unlock()
  return x
}

func main(args: [String]) {
  let lockObj = Lock()

  // Simple lock
  print("before lock")
  lock(lockObj) {
    print("    currently locked")
  }
  print("after lock")
}

main([])

File: test-src/22_lambdaLock.kt
class Lock() {
  fun lock() {
    println("  locked")
  }
  fun unlock() {
    println("  unlocked")
  }
}

fun <T> lock(lock: Lock, body: () -> T): T {
  lock.lock()
  val x = body()
  lock.unlock()
  return x
}

fun main(args: Array<String>) {
  val lockObj = Lock()

  // Simple lock
  println("before lock")
  lock(lockObj) {
    println("    currently locked")
  }
  println("after lock")
}
